File: 238-product-of-array-except-self/product-of-array-except-self.py
import math as mt

class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        ans = []
        n = len(nums)
        mul = 1

        for i in nums:
            mul *= i
            
        for i in range(n):
            if(nums[i] == 0):
                ans.append(mt.prod(nums[:i] + nums[i + 1:]))
            else: 
                ans.append(int(mul/nums[i]))

        return ans

File: 238-product-of-array-except-self/product-of-array-except-self.cpp
#define ll long long
#define vi vector<int>
#define vll vector<ll>
#define vs vector<string>
#define mpi map<int, int>
#define mps map<string, int>
#define pb push_back
#define all(x) (x).begin(), (x).end()

const int MOD = 1e9 + 7;

class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size();

        vi pre(n, 1);
        vi suf(n, 1);
        vi ans(n);

        for (int i = 1; i < n; i++) {
            pre[i] = pre[i - 1] * nums[i - 1];
        }

        for (int i = n - 2; i >= 0; i--) {
            suf[i] = suf[i + 1] * nums[i + 1];
        }

        for (int i = 0; i < n; i++) {
            ans[i] = pre[i] * suf[i];
        }

        return ans;
    }
};